### ISSUE TITLE
Moving objects inside and between lists

### ISSUE THREAD

**Issue by luleyleo** [automerge/automerge-rs#352]
Would it be possible to add a function to move an object to another object or move and object within a list?

I would like to change objects positions in the list when reordering them in the app, and also move them between multiple lists.

Is this something that could be implemented in this library, or would I have to work around it?

**Comment by ept** [comment]
Hi @luleyleo, this is something we've been thinking about, and we have published algorithms for [reordering lists](https://martin.kleppmann.com/2020/04/27/papoc-list-move.html) and [moving objects around the tree](https://martin.kleppmann.com/2021/10/07/crdt-tree-move-operation.html), but so far not yet had a chance to implement them in Automerge. If you (or anybody else) wants to have a go at implementing this, I can help get you started.

As a workaround for now, you can delete the object from the old position and re-add it to the new position. However, this has two problems:

* If two users concurrently move the same object, it will get duplicated.
* If another user concurrently makes a change to the object you're moving, that change will get discarded.

**Comment by luleyleo** [comment]
Thanks for the answer!

For now, I went with re-adding, but moving is a common operation for my app, so I would like to do it properly. I'll check out the algorithms you mentioned and see if I can help with it.

**Comment by bogiebro** [comment]
Is there still interest in adding this feature? I might be interested in taking it on.

**Comment by uinz** [comment]
If you wish to move elements in the list, consider using fractional indexing on the current state

---
- https://www.figma.com/blog/realtime-editing-of-ordered-sequences/#fractional-indexing
- https://github.com/rocicorp/fractional-indexing

**Comment by ept** [comment]
There is a work-in-progress implementation of a move operation by @LiangrunDa in #706, which we are hoping to make ready to merge soon.

**Comment by OlivierLDff** [comment]
That would be a very nice feature to have!

**Comment by tdomhan** [comment]
Ah nice! Would be great to have the ability to move things. Does the PR allow moves anywhere within the document?

**Comment by LiangrunDa** [comment]
Yes, the new algorithm allows moving element anywhere within the document!

**Comment by tdomhan** [comment]
Nice! thanks for the update. Looking forward to that feature.

**Comment by kitamura-tetsuo** [comment]
I managed to implement a working PoC (Proof of Concept) for move operations using Anthropic's Fable 5. 
You can check out the generated implementation here:
https://github.com/kitamura-tetsuo/automerge/pull/1

Since the resulting PR is quite large, I won't open a formal pull request here to avoid clustering the upstream repository. Instead, I wanted to share this as a reference. I believe the core maintainers could leverage AI agents like Fable 5 to implement this feature in a cleaner, more integrated way.

For reference, here is the prompt I used (along with attaching the paper `https://arxiv.org/pdf/2311.14007`):

> "Please implement move operations by referencing https://github.com/automerge/automerge/pull/706 and the attached paper."
